Fix: Achievement Refactoring

CuriousMagpie opened the pull request:
Fix for #11541 and #11257

### **Changes**
**Pet and Mount Color Achievements**
- Moved pet and mount color achievement key data from `notifications.vue` to `achievements.js` and created Objects `petColorAchievs` and `mountColorAchievs`
- Removed old pet and mount color achievement constants and key data from `userNotifications.js` and `notifications.vue`, replacing them with `ACHIEVEMENT_PET_COLOR` and `ACHIEVEMENT_MOUNT_COLOR`

**Animal Set Achievements**
- Moved animal set achievement key data from `notifications.vue` to `achievements.js` and created Objects `animalSetAchievs`
- Removed individual animal set achievement constants and key data from `userNotifications.js` and `notifications.vue`, replacing them with `ACHIEVEMENT_ANIMAL_SET`

**Beast Master/Mount Master/Triad Bingo Achievements**
- Changed name of this group of achievements from "Master" to "Stable" for clarity
- Created new constant file, `stableAchievements.js` and added it to `common/script/content/constants/index.js` and `common/script/content/index.js`
- Removed stable achievement constants and key data from `userNotifications.js` and `notifications.vue`, replacing them with `ACHIEVEMENT_STABLE`
- Updated `hooks.js` with new parameters in `beastMasterProgress`, `mountMasterProgress`, and `triadBingoProgress`

**Quest Series Achievements**
- Moved quest series achievement key data from `notifications.vue` to `achievements.js` and created Objects `questSeriesAchievs`
- Removed individual quest series achievement constants and key data from `userNotifications.js` and `notifications.vue`, replacing them with `ACHIEVEMENT_QUESTS`
- Updated `_.forEach(questSeriesAchievements` to call `ACHIEVEMENT_QUESTS` and removed snake case from modal
- Deleted `lostMasterclasser.vue`, `justAddWater.vue`, and `mindOverMatter.vue` from `components` as these files were not being used

**General Updates**
- Standardized achievement names across codebase, eg, `ULTIMATE_GEAR_ACHIEVEMENT` is now `ACHIEVEMENT_ULTIMATE_GEAR`

---
UUID: f4e5c6da-0617-48bf-b3bd-9f97636774a8

SabreCat committed:
WIP(achievements): updates from CuriousMagpie

CuriousMagpie committed:
Standardized achievement names across codebase & reordered notifications

CuriousMagpie committed:
first pass at pet color refactoring

CuriousMagpie committed:
work on ACHIEVEMENT_PET_COLOR modal text

CuriousMagpie committed:
work on ACHIEVEMENT_MOUNT_COLOR modal text

CuriousMagpie committed:
remove no longer used achievement constants from userNotification.js

CuriousMagpie committed:
refactored animal set achievements

CuriousMagpie committed:
moved quest achievements back under basicAchievs in achievements.js

CuriousMagpie committed:
started work on refactoring stable master achievements

CuriousMagpie committed:
renamed master achievements to stable achievements

CuriousMagpie committed:
fix typos

CuriousMagpie committed:
fix another typo

CuriousMagpie committed:
updated server/models/user/hooks.js

CuriousMagpie committed:
updated notifications.vue & hooks.js to use data from stableAcheivs obj

CuriousMagpie committed:
removed old achievement info from NOTIFICATIONS

CuriousMagpie committed:
cleaning up unused code/comments

CuriousMagpie committed:
finishing up stable achievements

CuriousMagpie committed:
fixed test/api/unit/models/user.test.js

CuriousMagpie committed:
started on quest series achievements

CuriousMagpie committed:
Merge remote-tracking branch 'upstream/develop' into cheevo-refactors

CuriousMagpie committed:
trying to trace questAchievementUpdate info

CuriousMagpie committed:
Merge branch 'develop' into cheevo-refactors

CuriousMagpie committed:
quest achievement progress, still not working

CuriousMagpie committed:
finished quest refactoring, deleted unused Vue files

CuriousMagpie committed:
fix lint errors

CuriousMagpie committed:
Merge remote-tracking branch 'upstream/develop' into cheevo-refactors

SabreCat commented on test/api/unit/models/user.test.js:
```diff
@@ -683,9 +683,15 @@ describe('User Model', () => {
+    // )).to.exist;
+      // expect(user.notifications.find(
+      // notification => notification.type === 'ACHIEVEMENT_TRIAD_BINGO',
+    // )).to.exist;
```
can remove these residual comments

CuriousMagpie replied:
done!

SabreCat commented on website/client/src/components/header/notificationsDropdown.vue:
```diff
@@ -167,9 +167,9 @@ export default {
+    // ACHIEVEMENT_JUST_ADD_WATER,
     ACHIEVEMENT_LOST_MASTERCLASSER,
-    ACHIEVEMENT_MIND_OVER_MATTER,
+    // ACHIEVEMENT_MIND_OVER_MATTER,
```
Same here, these can be deletions rather than comments

CuriousMagpie replied:
deleted!

SabreCat commented on website/client/src/components/notifications.vue:
```diff
@@ -227,219 +185,94 @@ const NOTIFICATIONS = {
+  //     achievement: 'seasonalSpecialist',
+  //   },
+  // },
+  // animal set collection achievements
```
more junk comments!

CuriousMagpie replied:
deleted!

SabreCat commented on website/client/src/components/notifications.vue:
```diff
@@ -140,82 +137,43 @@ import streak from './achievements/streak';
-  },
-  ACHIEVEMENT_BACK_TO_BASICS: {
+  // achievement notifications
+  ACHIEVEMENT: { // data filled in handleUserNotifications
```
If we're starting to pass the necessary information through via `data` in the notification object, can we go all the way down to this as the notification type for all achievements? Do we still need to split it out into several at all?

CuriousMagpie replied:
I didn't update these because they're related to the orientation/tutorial for new users, but certainly something I can work on! Do you want it as part of this PR or a future one?

CuriousMagpie replied:
Do you mean collapsing all the achievement types down into one item called `ACHIEVEMENT`: instead of `ACHIEVEMENT_CHALLENGE_JOINED`, `ACHIEVEMENT PARTY_ON`, etc?

SabreCat commented on website/client/src/components/notifications.vue:
```diff
@@ -893,55 +745,69 @@ export default {
+            this.showNotificationWithModal(notification);
+            Vue.set(this.user.achievements, achievement, true);
+            break;
+          }
```
continuing from above, see how these are all still almost identical? hmm...

I may poke at this if you're not feeling up to it!

CuriousMagpie replied:
Yeah, I see what you're getting at--and I can see how we could collapse these down into one--I'll give this one a try and see how it goes. I'm not quite seeing how it would work in the code above where we're defining the different achievements as they're pulling in data from different constant files `stableAchievs`, `questSeriesAchievs`, `animalSetAchievs` etc--part of the goal here was to make it easier to maintain each type of achievement instead of having one big list.

CuriousMagpie replied:
Hmmm. Now that I look at this more closely, I don't think I can collapse this, as it's pulling the achievement strings from each of the constant files... Let's discuss this.

SabreCat commented on website/server/models/group.js:
```diff
@@ -9,32 +9,23 @@ import { // eslint-disable-line import/no-cycle
 import { sendNotification as sendPushNotification } from '../libs/pushNotifications'; // eslint-disable-line import/no-cycle
-import { // eslint-disable-line import/no-cycle
+import {
+  // eslint-disable-line import/no-cycle
```
does this actually work? o.O I think it's more readable if we keep it on the same line as the `import` at least...

CuriousMagpie replied:
reverted back to `develop`

SabreCat commented on website/server/models/group.js:
```diff
@@ -1708,8 +1846,9 @@ schema.methods.checkChatSpam = function groupCheckChatSpam (user) {
-    && (!plan.dateTerminated || moment(plan.dateTerminated).isAfter(now));
+  return (
+    plan && plan.customerId && (!plan.dateTerminated || moment(plan.dateTerminated).isAfter(now))
+  );
```
this looks like an awful lot of irrelevant code style updates--is any of this relevant to achievements? If not, we can probably skip this file. If so, it might work better to just include the achievement-related pieces.

CuriousMagpie replied:
reverted to `develop`

SabreCat commented on website/server/models/userNotification.js:
```diff
@@ -5,73 +5,52 @@ import validator from 'validator';
-  'ACHIEVEMENT_WILD_BLUE_YONDER',
-  'ACHIEVEMENT_DOMESTICATED',
-  'ACHIEVEMENT_SHADY_CUSTOMER',
-  'ACHIEVEMENT_SHADE_OF_IT_ALL',
```
ok, this is going to be a problem for anyone who happens to have some legacy data. It might be these notifications are too ephemeral to stick around, but have you queried the db to see if any of these are still out there? If they are, we can mark these as deprecated, but deleting them entirely will lead to `enum` errors.

CuriousMagpie replied:
added back in with ` // deprecated`

SabreCat reviewed (changes requested):
A few questions! Do what cleanup you feel you can handle, then I can take it the rest of the way across the finish line.

CuriousMagpie committed:
Merge remote-tracking branch 'upstream/develop' into cheevo-refactors

CuriousMagpie committed:
Merge remote-tracking branch 'upstream/develop' into cheevo-refactors

CuriousMagpie committed:
cleaned up per change request; discussion needed on notifications.vue

CuriousMagpie committed:
Merge remote-tracking branch 'upstream/develop' into cheevo-refactors

CuriousMagpie committed:
added Zodiac Zookeeper achievement

CuriousMagpie committed:
added Birds of a Feather achievement

SabreCat committed:
Merge branch 'develop' into cheevo-refactors

SabreCat committed:
Merge branch 'develop' into cheevo-refactors

SabreCat committed:
WIP(achievements): refactor down to generic ACHIEVEMENT notification
Also moves the several achievement-related files from "constants" to an 
"achievements" folder in the content tree.

SabreCat commented:
Rolling this to Draft while I finish it up. Remaining to do:

- [ ] Add `data` object to the various places we `addNotification('ACHIEVEMENT')`
- - Make sure to call specific modals instead of the generic, where available (e.g. Rebirth)
- [ ] Handle opt-outs for streak and levelup achievements
- [ ] Update tests that look for ACHIEVEMENT notifications to also expect achievement-appropriate data

SabreCat committed:
fix(lint): comma oopsie

SabreCat committed:
fix(lint): argh another comma

SabreCat commented:
Merged in 1a74d2b3b04eb39660c47c64e41715e2f4289513 !
